README.md
# Lua Telemetry Flight Status for INAV/Taranis - v1.4.2

### FrSky SmartPort(S.Port), D-series, and F.Port telemetry on Taranis QX7, X-Lite, X9D, X9D+ and X9E transmitters

## Screenshots

#### Classic view on Taranis QX7 and X-Lite

![sample](assets/iNavQX71.png "Launch/pilot-based model orientation and location indicators")&nbsp;&nbsp;
![sample](assets/iNavQX72.png "Compass-based direction indicator")

#### Classic view on Taranis X9D, X9D+ and X9E

![sample](assets/iNavX9D.png "Classic view on Taranis X9D, X9D+ and X9E")

#### Pilot view for fixed wing pilots

![sample](assets/iNavQX7pilot.png "Pilot view on QX7 and X-Lite")&nbsp;&nbsp;
![sample](assets/iNavX9Dpilot.png "Pilot view on Taranis X9D, X9D+ and X9E")

## Features

* Works with all FrSky telemetry receivers (X-series, R9 series and D-series) and all FrSky Taranis transmitters
* Launch/pilot-based model orientation and location indicators (great for lost orientation/losing sight of your model)
* Compass-based direction indicator (with compass on multirotor or fixed-wing with GPS)
* Pilot view (glass cockpit) which includes attitude indicator as well as pilot-familiar layout of additional data
* Bar gauges for Fuel (% battery mAh capacity remaining), Battery voltage, RSSI strength, Transmitter battery, GPS accuracy (HDOP), Variometer (and Altitude for X9D, X9D+ and X9E transmitters)
* Display and voice alerts for flight modes and flight mode modifiers (altitude hold, heading hold, home reset, etc.)
* Voice notifications for % battery remaining (based on current), voltage low/critical, high altitude, lost GPS, ready to arm, armed, disarmed, etc.
* GPS info: Satellites locked, GPS accuracy (HDOP), GPS altitude, GPS coordinates. Also logs the last GPS location (reviewed from the config menu)
* If VTx control is desired, use [Taranis VTx](https://github.com/teckel12/Taranis-VTx) which uses less memory and allows for Lua Telemetry and VTx scripts to run together
* Display of current/maximum: Altitude, Distance, Speed and Current
* Display of current/minimum: Battery voltage, RSSI strength
* Title display of model name, flight timer, transmitter voltage and receiver voltage
* Menu configuration options can be changed from inside the script
* Speed and distance values are displayed in metric or imperial based on transmitter's telemetry settings

## Requirements

* [OpenTX v2.2.0+](http://www.open-tx.org/) running on Taranis QX7, X9D, X9D+, X9E or X-Lite (with OpenTX v2.2.2+)
* FkSky X-series, R9 series or D-series telemetry receiver: X4RSB, X8R, XSR, R-XSR, XSR-M, XSR-E, R9, R9 Slim, R9M, D8R-II plus, D8R-XP, D4R-II, etc.
* [INAV v1.7.3+](https://github.com/iNavFlight/inav/releases) running on your flight controller (INAV v2.0+ is suggested for full functionality)
* GPS - If you're looking for a GPS module, I suggest the [Beitian BN-880](https://www.banggood.com/UBLOX-NEO-M8N-BN-880-Flight-Control-GPS-Module-Dual-Module-Compass-p-971082.html)

## Suggested Sensors

* Altimeter/barometer (GPS altitude used if barometer not present)
* Magnetometer/compass for multi-rotor (fixed-wing craft use GPS for directional info)
* Current/amperage (for fuel gauge)

## Notes

* INAV v2.0+ is required for FkSky D-series telemetry and proper GPS accuracy (HDOP) display
* If using pilot view and INAV v2.0+, set `frsky_pitch_roll = ON` in CLI settings for more accurate attitude display
* INAV v1.9.1+ is required for F.Port compatibility
* INAV v1.8+ is required for `Home reset` voice notification
* OpenTX v2.2.2 is required for compatibility with Taranis X-Lite transmitter
* Crossfire is **NOT** supported due to missing telemetry data required by Lua Telemetry
    * Lua Telemetry can't fix this, Crossfire has missing flight mode info Lua Telemetry requires
* FrSky Horus transmitters are not currently supported due to the missing draw flags: `ERASE` and `FORCED`

## Setup

* [Lua Telemetry Wiki](https://github.com/iNavFlight/LuaTelemetry/wiki)
* [Download latest release](https://github.com/iNavFlight/LuaTelemetry/releases/latest)
* [Installation Instructions](https://github.com/iNavFlight/LuaTelemetry/wiki/Installation)
* [Upgrade Instructions](https://github.com/iNavFlight/LuaTelemetry/wiki/Upgrade)

## Information & Settings

* [Screen Description](https://github.com/iNavFlight/LuaTelemetry/wiki/Screen-Description)
* [Configuration Settings](https://github.com/iNavFlight/LuaTelemetry/wiki/Configuration-Settings)
* [Suggested Battery Settings](https://github.com/iNavFlight/LuaTelemetry/wiki/Suggested-Battery-Settings)
* [Change Log - Release History](https://github.com/iNavFlight/LuaTelemetry/wiki/Change-Log)

## Support

* [Tips & Common Problems](https://github.com/iNavFlight/LuaTelemetry/wiki/Tips-&-Common-Problems)
* [Support Issues](https://github.com/iNavFlight/LuaTelemetry/issues?q=is%3Aissue)

## Other

* [Upgrade to Development Build](https://github.com/iNavFlight/LuaTelemetry/wiki/Upgrade-to-Development-Build)
* [License](https://github.com/iNavFlight/LuaTelemetry/blob/master/LICENSE)

CHANGES.md
# Release History

## v1.4.0 - 08/21/2018

* Pilot view (glass cockpit) which includes attitude indicator as well as pilot-familiar layout of additional data
* Shows GPS fix accuracy (HDOP) as strength indicator graph or decimal value (required INAV 2.0.0)
* You can customize when weak GPS fix accuracy (HDOP) triggers alert
* Allows speed sensor selection between GPS speed or pitot sensor's air speed (if available)
* GPS coordinates and altitude are displayed before launch even if there's no GPS fix
* Fuel can report mAh or mWh used instead of percent fuel remaining (percent highly suggested for fuel level alerts)
* The setup, config menu and views loads as separate scripts, greatly reducing memory used
* Variometer graph now correctly uses vertical speed as source instead of Z-axis accelerometer
* Config option under Variometer to also report altitude as voice notifications
* X9D display now uses gray when appropriate to enhance view clarity
* Disable config setting (and set to default) if sensor isn't present
* Config menu wraps from top to bottom (and vice versa) to more easily locate desired config option, also remembers last menu option
* Config menu supports button hold-repeat on X9D and X9D+ to make changes easier
* Config menu fuel percentage warning and critical changed to 1% steps to increase flexibility
* Only shows the last good GPS coordinates instead of the last 5 to reduce complexity (didn't provide more useful info either)
* Removed `getLastPos()` function dependency which is faster, cleans up code and could allow Horus support
* Fixed an issue with "NO TELEM" and "THR WARN" being displayed incorrectly
* Cleaned up readme, broke up into Wiki pages

## v1.3.1 - 05/29/2018

* Support added for the X-Lite transmitter (requires OpenTX v2.2.2)
* Option to set the low and critical warning level for fuel percentage remaining
* Option to display the transmitter voltage as a graph and/or the numerical value
* Modified how battery cell count is calculated
* Can now set cell critical voltage to as low as 2.6v to support 18650 batteries
* Slightly adjusted display positioning to show greater GPS accuracy on QX7 and X-Lite transmitters

## v1.3.0 - 05/13/2018

* Compatibility with new F.Port protocol (requires INAV v1.9.1+)
* Support added for D-series receivers with telemetry (requires INAV v2.0+)
* Option to show GPS coordinates as degrees and minutes or in geocoding format instead of the previous decimal only
* Shows `THR WARN` if not okay to fly and throttle isn't at minimum position
* If distance sensor name is the default `0420`, the max distance now displays correctly
* Updated data digits to display before truncating unit of measure

## v1.2.5 - 04/28/2018

* Uses GPS for launch-based altitude if barometer isn't present
* No longer requires `0420` sensor to be renamed `Dist`
* Better error handling, specially if `iNav` folder doesn't exist
* Renamed PASSTHRU mode MANUAL and add audio file for manual mode
* Correct altitude unit designation in configuration when barometer isn't present
* Adjustment to data display for min/max indicator

## v1.2.4 - 04/17/2018

* Satellite and altitude hold are now icons
* Added flashing notification for headfree mode on QX7 (already existed on X9D)
* Smarter unit designation resulting in a cleaner display (ft = ', MPH & km/h)

## v1.2.3 - 12/19/2017

* Logs the last 5 unique GPS coordinates
* Updated readme screenshots to show variometer
* Release history is now here instead of in readme file

## v1.2.2 - 12/09/2017

* Last 5 GPS coordinates can be reviewed from the config menu
* Resolved issue where if telemetry was lost/recovered it would incorrectly give voice alerts for flight modes/engines armed/disarmed
* Config menu can be accessed at any time, even when armed

## v1.2.1 - 12/01/2017

* Lots of new/changed config settings (press `Menu` button to access) - **Please review!**
* Enhanced and cleaned up config menu, centered menu on QX7 & X9D/+
* Variometer shows if model is gaining or decreasing altitude
* Cell voltage battery view shows with two digits of precision
* Added `S` (South) to X9D/+ directional display

## v1.2.0 - 11/18/2017

* Lua Telemetry is now pre-compiled to greatly reduce memory (source still available)
* Press `Menu` button (when not armed) to modify user configuration options
* Fixed `FAILSAFE`, `RTH`, `NOT OK` & `NO TELEM` modes to flash as they should
* Barometer and magnetometer are now optional (but suggested for full functionality)
* Headfree indication on QX7 changed to show directional indicators as dotted lines
* Startup message/version and error if not running OpenTX v2.2.0 or later

## v1.1.7 - 11/02/2017

* Fix for the default unit type of the `Dist` (`0420`) sensor

## v1.1.6 - 11/01/2017

* On home reset, reset GPS home position, orientation and distance
* Option to display average battery cell voltage instead of total voltage
* Extra digit for data on X9D & X9D+ transmitters
* Variable cleanup saving memory

## v1.1.5 - 10/20/2017

* Voice notification for `Home Reset` with INAV v1.8+
* Moved head free warning on QX7 to top center
* Values convert from decimal to integer when larger to allow for more room
* Better text centering and right justification technique
* Cleaned up code saving more memory

## v1.1.4 - 10/13/2017

* More accurate max altitude alerts and altitude flashes when above max altitude
* Long-press <Enter> resets values (suggest doing this between flights before armed)

## v1.1.3 - 10/10/2017

* Shows metric or imperial values based on transmitter telemetry settings

## v1.1.2 - 10/06/2017

* Lots of refactoring which greatly reduced memory usage
* Re-enabled altitude bar gauge for X9D, X9D+ & X9E transmitters
* Better information layout if no current sensor is present
* Refactored GPS lock calculation to prevent script syntax errors

## v1.1.1 - 09/28/2017

* Refactored code to reduce memory
* Removed altitude bar gauge for X9D, X9D+ & X9E transmitters (used too much memory?)


src/iNav.lua
-- Lua Telemetry Flight Status for INAV/Taranis
-- Author: https://github.com/teckel12
-- Docs: https://github.com/iNavFlight/LuaTelemetry

local VERSION = "1.4.2"
local FILE_PATH = "/SCRIPTS/TELEMETRY/iNav/"
local FLASH = 3
local SMLCD = LCD_W < 212
local tmp, view

local config = loadScript(FILE_PATH .. "config", "bT")(SMLCD)
collectgarbage()

local modes, units = loadScript(FILE_PATH .. "modes", "bT")(FLASH)
local configCnt = loadScript(FILE_PATH .. "load", "bT")(config, FILE_PATH)
collectgarbage()

local data, PREV, INCR, NEXT, DECR, MENU = loadScript(FILE_PATH .. "data", "bT")()
collectgarbage()

loadScript(FILE_PATH .. "reset", "bT")(data)
loadScript(FILE_PATH .. "other", "bT")(config, data, units, FILE_PATH)
collectgarbage()

local function playAudio(f, a)
	if config[4].v == 2 or (config[4].v == 1 and a ~= nil) then
		playFile(FILE_PATH .. f .. ".wav")
	end
end

local function gpsDegMin(c, lat)
	local gpsD = math.floor(math.abs(c))
	return gpsD .. string.format("\64%05.2f", (math.abs(c) - gpsD) * 60) .. (lat and (c >= 0 and "N" or "S") or (c >= 0 and "E" or "W"))
end

local function gpsIcon(x, y)
	lcd.drawLine(x + 1, y, x + 5, y + 4, SOLID, 0)
	lcd.drawLine(x + 1, y + 1, x + 4, y + 4, SOLID, 0)
	lcd.drawLine(x + 1, y + 2, x + 3, y + 4, SOLID, 0)
	lcd.drawLine(x, y + 5, x + 2, y + 5, SOLID, 0)
	lcd.drawPoint(x + 4, y + 1)
	lcd.drawPoint(x + 1, y + 4)
end

local function lockIcon(x, y)
	lcd.drawFilledRectangle(x, y + 2, 5, 4, 0)
	lcd.drawLine(x + 1, y, x + 3, y, SOLID, 0)
	lcd.drawPoint(x + 1, y + 1)
	lcd.drawPoint(x + 3, y + 1)
	lcd.drawPoint(x + 2, y + 3, ERASE)
end

local function homeIcon(x, y)
	lcd.drawPoint(x + 3, y - 1)
	lcd.drawLine(x + 2, y, x + 4, y, SOLID, 0)
	lcd.drawLine(x + 1, y + 1, x + 5, y + 1, SOLID, 0)
	lcd.drawLine(x, y + 2, x + 6, y + 2, SOLID, 0)
	lcd.drawLine(x + 1, y + 3, x + 1, y + 5, SOLID, 0)
	lcd.drawLine(x + 5, y + 3, x + 5, y + 5, SOLID, 0)
	lcd.drawLine(x + 2, y + 5, x + 4, y + 5, SOLID, 0)
	lcd.drawPoint(x + 3, y + 4)
end

local function hdopGraph(x, y, s)
	local tmp = ((data.armed or data.modeId == 6) and data.hdop < 11 - config[21].v * 2) or not data.telemetry
	if config[22].v == 0 then
		if tmp then
			lcd.drawText(x, y, "    ", SMLSIZE + FLASH)
		end
		for i = 4, 9 do
			lcd.drawLine(x - 8 + (i * 2), (data.hdop >= i or not SMLCD) and y + 8 - i or y + 5, x - 8 + (i * 2), y + 5, SOLID, (data.hdop >= i or SMLCD) and 0 or GREY_DEFAULT)
		end
	else
		lcd.drawText(x + 12, s == SMLSIZE and y or y - 2, (data.hdop == 0 and not data.gpsFix) and "--" or (9 - data.hdop) / 2 + 0.8, s + RIGHT + (tmp and FLASH or 0))
	end
end

local function background()
	data.rssi = getValue(data.rssi_id)
	if data.rssi > 0 then
		data.telemetry = true
		data.telemFlags = 0
		data.mode = getValue(data.mode_id)
		data.rxBatt = getValue(data.rxBatt_id)
		data.satellites = getValue(data.satellites_id)
		data.gpsAlt = data.satellites > 1000 and getValue(data.gpsAlt_id) or 0
		data.heading = getValue(data.heading_id)
		data.altitude = getValue(data.altitude_id)
		if data.altitude_id == -1 and data.gpsAltBase and data.gpsFix and data.satellites > 3000 then
			data.altitude = data.gpsAlt - data.gpsAltBase
		end
		data.distance = getValue(data.distance_id)
		data.speed = getValue(data.speed_id)
		if data.showCurr then
			data.current = getValue(data.current_id)
			data.currentMax = getValue(data.currentMax_id)
			data.fuel = getValue(data.fuel_id)
		end
		data.altitudeMax = getValue(data.altitudeMax_id)
		data.distanceMax = getValue(data.distanceMax_id)
		data.speedMax = getValue(data.speedMax_id)
		data.batt = getValue(data.batt_id)
		data.battMin = getValue(data.battMin_id)
		data.cells = (data.batt / data.cells > 4.3) and math.floor(data.batt / 4.3) + 1 or data.cells
		data.cell = data.batt / data.cells
		data.cellMin = data.battMin / data.cells
		data.rssiMin = getValue(data.rssiMin_id)
		data.vspeed = getValue(data.vspeed_id)
		if data.pitchRoll then
			data.pitch = getValue(data.pitch_id)
			data.roll = getValue(data.roll_id)
		else
			data.accx = getValue(data.accx_id)
			data.accy = getValue(data.accy_id)
			data.accz = getValue(data.accz_id)
		end
		data.rssiLast = data.rssi
		local gpsTemp = getValue(data.gpsLatLon_id)
		if type(gpsTemp) == "table" and gpsTemp.lat ~= nil and gpsTemp.lon ~= nil then
			data.gpsLatLon = gpsTemp
			if data.satellites > 1000 and gpsTemp.lat ~= 0 and gpsTemp.lon ~= 0 then
				data.gpsFix = true
				config[15].l[0] = gpsTemp
			end
		end
		-- Dist doesn't have a known unit so the transmitter doesn't auto-convert
		if data.distance_unit == 10 then
			data.distance = math.floor(data.distance * 3.28084 + 0.5)
			data.distanceMax = data.distanceMax * 3.28084
		end
		if data.distance > 0 then
			data.distanceLast = data.distance
		end
	else
		data.telemetry = false
		data.telemFlags = FLASH
	end
	data.txBatt = getValue(data.txBatt_id)
	data.throttle = getValue(data.throttle_id)

	local armedPrev = data.armed
	local headFreePrev = data.headFree
	local headingHoldPrev = data.headingHold
	local altHoldPrev = data.altHold
	local homeReset = false
	local modeIdPrev = data.modeId
	local preArmMode = false
	data.modeId = 1 -- No telemetry
	if data.telemetry then
		data.armed = false
		data.headFree = false
		data.headingHold = false
		data.altHold = false
		local modeA = data.mode / 10000
		local modeB = data.mode / 1000 % 10
		local modeC = data.mode / 100 % 10
		local modeD = data.mode / 10 % 10
		local modeE = data.mode % 10
		if bit32.band(modeD, 2) == 2 then
			data.modeId = 2 -- Horizon
		elseif bit32.band(modeD, 1) == 1 then
			data.modeId = 3 -- Angle
		else
			data.modeId = 4 -- Acro
		end
		data.headFree = bit32.band(modeB, 4) == 4 and true or false
		data.headingHold = bit32.band(modeC, 1) == 1 and true or false
		if bit32.band(modeE, 4) == 4 then
			data.armed = true
			data.altHold = (bit32.band(modeC, 2) == 2 or bit32.band(modeC, 4) == 4) and true or false
			homeReset = data.satellites >= 4000 and true or false
			data.modeId = bit32.band(modeC, 4) == 4 and 7 or data.modeId -- pos hold
		else
			preArmMode = data.modeId
			data.modeId = (bit32.band(modeE, 2) == 2 or modeE == 0) and (data.throttle > -1000 and 12 or 5) or 6 -- Not OK to arm(5) / Throttle warning(12) / Ready to fly(6)
		end
		if bit32.band(modeA, 4) == 4 then
			data.modeId = 11 -- Failsafe
		elseif bit32.band(modeB, 1) == 1 then
			data.modeId = 10 -- RTH
		elseif bit32.band(modeD, 4) == 4 then
			data.modeId = 9 -- Manual
		elseif bit32.band(modeB, 2) == 2 then
			data.modeId = 8 -- Waypoint
		elseif bit32.band(modeB, 8) == 8 then
			data.modeId = 13 -- Cruise
		end
	end

	-- Voice alerts
	local vibrate = false
	local beep = false
	if data.armed and not armedPrev then -- Engines armed
		data.timerStart = getTime()
		data.headingRef = data.heading
		data.gpsHome = false
		data.battPercentPlayed = 100
		data.battLow = false
		data.showMax = false
		data.showDir = false
		data.configStatus = 0
		if not data.gpsAltBase and data.gpsFix then
			data.gpsAltBase = data.gpsAlt
		end
		playAudio("engarm", 1)
	elseif not data.armed and armedPrev then -- Engines disarmed
		if data.distanceLast <= data.distRef then
			data.headingRef = -1
			data.showDir = true
			data.gpsAltBase = false
		end
		playAudio("engdrm", 1)
	end
	if data.gpsFix ~= data.gpsFixPrev then -- GPS status change
		playAudio("gps", not data.gpsFix and 1 or nil)
		playAudio(data.gpsFix and "good" or "lost", not data.gpsFix and 1 or nil)
	end
	if modeIdPrev ~= data.modeId then -- New flight mode
		if data.armed and modes[data.modeId].w ~= nil then
			playAudio(modes[data.modeId].w, modes[data.modeId].f > 0 and 1 or nil)
		elseif not data.armed and data.modeId == 6 and modeIdPrev == 5 then
			playAudio(modes[data.modeId].w)
		end
	elseif preArmMode ~= false and data.preArmModePrev ~= preArmMode then
		playAudio(modes[preArmMode].w)
	end
	data.hdop = math.floor(data.satellites / 100) % 10
	if data.headingHold ~= headingHoldPrev then -- Heading hold status change
		playAudio("hedhld")
		playAudio(data.headingHold and "active" or "off")
	end
	if data.headFree ~= headFreePrev then -- Head free status change
		playAudio(data.headFree and "hfact" or "hfoff", 1)
	end
	if data.armed then
		data.distanceLast = data.distance
		if config[13].v == 1 then
			data.timer = (getTime() - data.timerStart) / 100 -- Armed so update timer
		elseif config[13].v > 1 then
			data.timer = model.getTimer(config[13].v - 2)["value"]
		end
		if data.altHold ~= altHoldPrev then -- Alt hold status change
			playAudio("althld")
			playAudio(data.altHold and "active" or "off")
		end
		if homeReset and not data.homeResetPrev then -- Home reset
			playAudio("homrst")
			data.gpsHome = false
			data.headingRef = data.heading
		end
		if data.altitude + 0.5 >= config[6].v and config[12].v > 0 then -- Altitude alert
			if getTime() > data.altNextPlay then
				if config[4].v > 0 then
					playNumber(data.altitude + 0.5, data.altitude_unit)
				end
				data.altNextPlay = getTime() + 1000
			else
				beep = true
			end
		elseif config[7].v > 1 then -- Vario voice
			if math.abs(data.altitude - data.altLastAlt) + 0.5 >= config[24].l[config[24].v] then
				if math.abs(data.altitude + 0.5 - data.altLastAlt) / config[24].l[config[24].v] > 1.5 then
					tmp = math.floor((data.altitude + 0.5) / config[24].l[config[24].v]) * config[24].l[config[24].v]
				else
					tmp = math.floor(data.altitude / config[24].l[config[24].v] + 0.5) * config[24].l[config[24].v]
				end
				if tmp > 0 and getTime() > data.altNextPlay then
					playNumber(tmp, data.altitude_unit)
					data.altLastAlt = tmp
					data.altNextPlay = getTime() + 500
				end
			end
		end
		if config[23].v == 0 and data.battPercentPlayed > data.fuel and config[11].v == 2 and config[4].v == 2 then -- Fuel notifications
			if data.fuel >= config[17].v and data.fuel <= config[18].v and data.fuel > config[17].v then -- Fuel low
				playAudio("batlow")
				playNumber(data.fuel, 13)
				data.battPercentPlayed = data.fuel
			elseif data.fuel % 10 == 0 and data.fuel < 100 and data.fuel > config[18].v then -- Fuel 10% notification
				playAudio("battry")
				playNumber(data.fuel, 13)
				data.battPercentPlayed = data.fuel
			end
		end
		if ((config[23].v == 0 and data.fuel <= config[17].v) or data.cell < config[3].v) and config[11].v > 0 then -- Voltage/fuel critial
			if getTime() > data.battNextPlay then
				playAudio("batcrt", 1)
				if config[23].v == 0 and data.fuel <= config[17].v and data.battPercentPlayed > data.fuel and config[4].v > 0 then
					playNumber(data.fuel, 13)
					data.battPercentPlayed = data.fuel
				end
				data.battNextPlay = getTime() + 500
			else
				vibrate = true
				beep = true
			end
			data.battLow = true
		elseif data.cell < config[2].v and config[11].v == 2 then -- Voltage notification
			if not data.battLow then
				playAudio("batlow")
				data.battLow = true
			end
		else
			data.battNextPlay = 0
		end
		if (data.headFree and config[9].v == 1) or modes[data.modeId].f ~= 0 then
			if data.modeId ~= 10 or (data.modeId == 10 and config[8].v == 1) then
				beep = true
				vibrate = true
			end
		elseif data.rssi < data.rssiLow and config[10].v == 1 then
			if data.rssi < data.rssiCrit then
				vibrate = true
			end
			beep = true
		end
		if data.hdop < 11 - config[21].v * 2 then
			beep = true
		end
		if vibrate and (config[5].v == 1 or config[5].v == 3) then
			playHaptic(25, 3000)
		end
		if beep and config[5].v >= 2 then
			playTone(2000, 100, 3000, PLAY_NOW)
		end
	else
		data.battLow = false
		data.battPercentPlayed = 100
	end
	data.gpsFixPrev = data.gpsFix
	data.homeResetPrev = homeReset
	data.preArmModePrev = preArmMode

	if data.armed and data.gpsFix and data.gpsHome == false then
		data.gpsHome = data.gpsLatLon
	end
end

local function run(event)
	--[[ Lock display at 10fps
	if event == 0 then
		if getTime() < data.refresh then
			return 0
		end
		data.refresh = getTime() + 10
	end
	]]
	lcd.clear()

	-- Display system error
	if data.systemError then
		lcd.drawText((LCD_W - string.len(data.systemError) * 5.2) / 2, 27, data.systemError)
		return 0
	end

	-- Startup message
	if data.startup == 1 then
		data.startupTime = getTime()
		data.startup = 2
	elseif data.startup == 2 and getTime() - data.startupTime >= 200 then
		data.startup = 0
	end

	-- Config menu or views
	if data.configStatus == 0 and event == MENU then
		data.configStatus = data.configLast
	end
	if data.configStatus > 0 then
		if data.v ~= 9 then
			view = nil
			collectgarbage()
			--if data.simu then
				view = loadScript(FILE_PATH .. "menu", "bT")()
			--else
			--	view = loadfile(FILE_PATH .. "menu.luac")()
			--end
			data.v = 9
		end
		view(data, config, event, configCnt, gpsDegMin, FILE_PATH, SMLCD, FLASH, PREV, INCR, NEXT, DECR)
	else
		-- User input
		if not data.armed and data.configStatus == 0 then
			-- Toggle showing max/min values
			if event == PREV or event == NEXT then
				data.showMax = not data.showMax
			end
			-- Initalize variables on long <Enter>
			if event == EVT_ENTER_LONG then
				loadScript(FILE_PATH .. "reset", "bT")(data)
			end
		end
		if event == NEXT or event == PREV then
			data.showDir = not data.showDir
		end

		-- Views
		if data.v ~= config[25].v then
			view = nil
			collectgarbage()
			--if data.simu then
				view = loadScript(FILE_PATH .. (config[25].v == 1 and "pilot" or "view"), "bT")()
			--else
			--	view = loadfile(FILE_PATH .. (config[25].v == 1 and "pilot.luac" or "view.luac"))()
			--end
			data.v = config[25].v
		end
		view(data, config, modes, units, gpsDegMin, gpsIcon, lockIcon, homeIcon, hdopGraph, VERSION, SMLCD, FLASH, FILE_PATH)
	end
	collectgarbage()

	-- Title
	lcd.drawFilledRectangle(0, 0, LCD_W, 8, FORCE)
	lcd.drawText(0, 0, data.modelName, INVERS)
	if config[13].v > 0 then
		lcd.drawTimer(SMLCD and 60 or 150, 1, data.timer, SMLSIZE + INVERS)
	end
	if config[19].v > 0 then
		lcd.drawFilledRectangle(86, 1, 19, 6, ERASE)
		lcd.drawLine(105, 2, 105, 5, SOLID, ERASE)
		tmp = math.max(math.min((data.txBatt - data.txBattMin) / (data.txBattMax - data.txBattMin) * 17, 17), 0) + 86
		for i = 87, tmp, 2 do
			lcd.drawLine(i, 2, i, 5, SOLID, FORCE)
		end
	end
	if config[19].v ~= 1 then
		lcd.drawText(SMLCD and (config[14].v == 1 and 105 or LCD_W) or 128, 1, string.format("%.1f", data.txBatt) .. "V", SMLSIZE + RIGHT + INVERS)
	end
	if data.rxBatt > 0 and data.telemetry and config[14].v == 1 then
		lcd.drawText(LCD_W, 1, string.format("%.1f", data.rxBatt) .. "V", SMLSIZE + RIGHT + INVERS)
	end

	--[[ Show FPS
	data.frames = data.frames + 1
	lcd.drawText(SMLCD and 57 or 80, 1, string.format("%.1f", data.frames / (getTime() - data.fpsStart) * 100), SMLSIZE + RIGHT + INVERS)
	]]

	return 0
end

return { run = run, background = background }